camptix.php

```diff
@@ -2933,23 +2923,26 @@ public static function esc_csv( $fields ) {
 		);
 
 		foreach( $fields as $index => $field ) {
-			// Escape trigger characters at the start of a new field
-			$first_cell_character = mb_substr( $field, 0, 1 );
-			$is_trigger_character = in_array( $first_cell_character, $active_content_triggers, true );
-			$is_delimiter         = in_array( $first_cell_character, $delimiters,              true );
 
-			if ( $is_trigger_character || $is_delimiter ) {
-				$field = "'" . $field;
+			if ( ! is_string( $field ) ) {
+				continue;
 			}
 
-			// Escape trigger characters that follow delimiters
-			foreach ( $delimiters as $delimiter ) {
-				foreach ( $active_content_triggers as $trigger ) {
-					$field = str_replace( $delimiter . $trigger, $delimiter . "'" . $trigger, $field );
+			$escaped_field = '';
+
+			// Escape trigger characters that follow delimiters, or are at the start
+			$is_prev_char_delimiter = true;
+			for ( $i = 0; $i < strlen( $field ); $i++ ) {
```

Comment:
The previous code used `mb_substr()` to mitigate multibyte payloads, so we should probably use `mb_strlen()` here as well. IIRC that's not available on all PHP versions, but Core might have a stub for it. I think that's the case w/ `mb_substr()`.

We'd need to make sure that there aren't any problems with using a multibyte string as a character array in `$char = $field[ $i ];` . I don't think there would be if we use `mb_strlen()`, but don't know for sure.

Do we have any unit tests for multibyte chars? If not we should add some, and make sure they pass the current code as well as these changes. There are probably some public reports on Hacker1 that have example payloads. If not let me know and I might be able to find one.

Reply:
I have added some tests for mb strings, and looks like they are passing. I think so because none of the trigger chars or delimiters are multi-byte characters, but we have to revisit this if this assumption is not valid.

Reply:
Interestingly, using `mb_*`  reduce the performance for some reason? So I have not added them till I figure out why. But for now this should be good enough.

Reply:
How much does it reduce the performance? I'm guessing it wouldn't be a significant amount, so it feels safer to err on the side of caution.

If it is significant, I think we'd need to do some very thorough research and testing to make sure we wouldn't be introducing a vulnerability by leaving it out.